src/utils/waitForAll.ts

```diff
@@ -0,0 +1,65 @@
+import { Atom, atom } from 'jotai'
+
+import { getWeakCacheItem, setWeakCacheItem } from './weakCache'
+
+const waitForAllCache = new WeakMap()
+
+export function waitForAll<Values extends Record<string, unknown>>(
+  atoms: { [K in keyof Values]: Atom<Values[K]> }
+): Atom<Values>
+
+export function waitForAll<Values extends unknown[]>(
+  atoms: { [K in keyof Values]: Atom<Values[K]> }
+): Atom<Values>
+
+export function waitForAll<Values extends Record<string, unknown> | unknown[]>(
+  atoms: { [K in keyof Values]: Atom<Values[K]> }
+) {
+  const cachedAtom =
+    Array.isArray(atoms) && getWeakCacheItem(waitForAllCache, atoms)
+  if (cachedAtom) {
+    return cachedAtom as Atom<Values>
+  }
+  const derivedAtom = atom((get) => {
+    const promises: Promise<unknown>[] = []
+    const values = unwrapAtoms(atoms).map((anAtom, index) => {
+      try {
+        return get(anAtom)
+      } catch (e) {
+        if (e instanceof Promise) {
+          promises[index] = e
+        } else {
+          throw e
+        }
+      }
+    })
+    if (promises.length) {
+      throw Promise.all(promises)
```

Comment:
```suggestion
      return Promise.all(promises)
```
It would be nice if we can do this, because it's more idiomatic. (throwing promise is something we should hide as implementation details.